From c5ea555b0c835dd66daf0dfb37e6c7bffbcdc481 Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Tue, 18 Feb 2025 13:36:00 +0100
Subject: [PATCH 784/954] feat(stm32mp2): configure TAMP before reading TAMP_SR

TAMP RIF configuration is maintained by the backup domain.
Hence when switching from M33TD to A35TD without losing the backup
domain, TAMP registers might not be accessible.
To prevent an IAC, set configuration of TAMP_R0CIDCFGR if necessary
before reading TAMP_SR.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 71e0a87bb277ccfa9cc66d5c7b9010b195c6a051 feat(stm32mp2): configure TAMP before reading TAMP_SR]
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Change-Id: Ic626c5122a008716d1501a86116767f786e82014
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/437673
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c | 32 ++++++++++++++++++++++++++++++-
 plat/st/stm32mp2/stm32mp2_def.h   |  5 +++++
 2 files changed, 36 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index b64f563990..ecd5288eae 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -224,8 +224,26 @@ static void handle_confirmed_tamper(uint32_t bit_off __unused)
 	panic();
 }
 
+static bool tamper_is_accesssible(void)
+{
+	uint32_t r0cidcfgr = mmio_read_32(TAMP_BASE + TAMP_R0CIDCFGR);
+
+	return !(((r0cidcfgr & TAMP_R0CIDCFGR_CFEN) == TAMP_R0CIDCFGR_CFEN) &&
+		 ((r0cidcfgr & TAMP_R0CIDCFGR_CID) != TAMP_R0CIDCFGR_CID1));
+}
+
 static bool lse_tamper_detection(void)
 {
+	/*
+	 * In case the backup domain hasn't been reset on warm boot cases, the
+	 * TAMP RIF configuration is maintained.
+	 * Therefore, TAMP_SR may not be accessible. In this case skip tamper
+	 * detection.
+	 */
+	if (!tamper_is_accesssible()) {
+		return false;
+	}
+
 	if ((mmio_read_32(TAMP_BASE + TAMP_SR) & TAMP_SR_LSE_MONITORING) != 0U) {
 		mmio_clrbits_32(RCC_BASE + RCC_BDCR, RCC_BDCR_LSECSSON);
 		mmio_clrbits_32(RCC_BASE + RCC_BDCR, RCC_BDCR_LSEON);
@@ -279,8 +297,20 @@ static void reset_backup_domain(void)
 static void check_tamper_event(bool lse_tamper_occured)
 {
 #if !STM32MP_M33_TDCID
-	uint32_t sr = mmio_read_32(TAMP_BASE + TAMP_SR);
+	uint32_t sr;
+
+	/*
+	 * In case the backup domain hasn't been reset on warm boot cases, the
+	 * TAMP RIF configuration is maintained.
+	 * Therefore, TAMP_SR may not be accessible. In this case skip tamper
+	 * detection.
+	 */
+	if (!tamper_is_accesssible()) {
+		INFO("TAMP IP not accessible, skip tamper verification\n");
+		return;
+	}
 
+	sr = mmio_read_32(TAMP_BASE + TAMP_SR);
 	if (sr == 0U) {
 		return;
 	}
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 19cf8c9c0b..1f02a799a8 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -584,6 +584,11 @@ enum ddr_type {
 #define TAMP_BKP_SEC_NUMBER		U(10)
 #define TAMP_SCR			U(0x3C)
 #define TAMP_COUNTR			U(0x40)
+#define TAMP_R0CIDCFGR			U(0x80)
+#define TAMP_R0CIDCFGR_CFEN		BIT_32(0)
+#define TAMP_R0CIDCFGR_CID		GENMASK_32(6, 4)
+#define TAMP_R0CIDCFGR_CID_SHIFT	U(4)
+#define TAMP_R0CIDCFGR_CID1		(RIF_CID1 << TAMP_R0CIDCFGR_CID_SHIFT)
 
 #define TAMP_SR_LSE_MONITORING		BIT_32(18)
 
-- 
2.53.0

